pkg/kubelet/kubelet_pods.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
+// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
+// capability, the pod contains a privileged container, or the pod has a host path volume.
+func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
+	if hasPrivilegedContainer(pod) || hasHostNamespace(pod) ||
+		hasHostVolume(pod) || hasNonNamespacedCapability(pod) || kl.hasHostMountPVC(pod) {
+		return true
+	}
+	return false
+}
+
+// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
+func hasPrivilegedContainer(pod *api.Pod) bool {
```

Comment:
Ohhh!!! Does this mean that with userns enabled, containers in a pod will
automatically share userns with docker?

On Fri, Sep 2, 2016 at 2:19 PM, Mrunal Patel <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet_pods.go
> <https://github.com/kubernetes/kubernetes/pull/31169#discussion_r77413259>
> :
>
> > +	"k8s.io/kubernetes/pkg/api"
> > +)
> > +
> > +// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
> > +// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
> > +// capability, the pod contains a privileged container, or the pod has a host path volume.
> > +func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
> > +	if hasPrivilegedContainer(pod) || hasHostNamespace(pod) ||
> > +		hasHostVolume(pod) || hasNonNamespacedCapability(pod) || kl.hasHostMountPVC(pod) {
> > +		return true
> > +	}
> > +	return false
> > +}
> > +
> > +// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
> > +func hasPrivilegedContainer(pod *api.Pod) bool {
>
> AFAIK, we don't share userns in a pod. So why would it matter if a single
> container in a multi-container pod were to use native userns?
>
> Actually, it is required that a container shares a userns with another
> container if it shares any namespace with the other container else it won't
> have the correct capabilities in the namespace. (This happens by default in
> docker).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169/files/423ccec9b7b7582c3a4298015d4d23a71d664567#r77413259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKGUbkJJZLbPYiXhYhlFA9xq0AxoFks5qmJLMgaJpZM4JqWln>
> .
>